code/controllers/subsystems/overlays.dm

```diff
@@ -1,242 +1,239 @@
- * @param special If true, returns special overlays like emissives and em_blockers
- */
-/proc/get_overlays(atom/other, priority, special)
-	var/list/including = list()
-	if (!other)
-		return including
-	for (var/image/I as anything in other.our_overlays)
-		if (!special && I.plane > 0)
-			continue
-		including += I
-	if (!priority)
-		return including
-	for (var/image/I as anything in other.priority_overlays)
-		if (!special && I.plane > 0)
-			continue
-		including += I
-	return including
+SUBSYSTEM_DEF(overlays)
+	name = "Overlay"
+	flags = SS_TICKER
+	wait = 1 // SS_TICKER - Ticks
+	priority = FIRE_PRIORITY_OVERLAYS
+	init_order = INIT_ORDER_OVERLAY
+
+	/// The queue of atoms that need overlay updates.
+	var/static/tmp/list/queue = list()
+
+	/// A list([icon] = list([state] = [appearance], ...), ...) cache of appearances.
+	var/static/tmp/list/state_cache = list()
+
+	/// A list([icon] = [appearance], ...) cache of appearances.
+	var/static/tmp/list/icon_cache = list()
+
+	/// The number of appearances currently cached.
+	var/static/tmp/cache_size = 0
+
+
+/datum/controller/subsystem/overlays/Recover()
+	queue.Cut()
+	state_cache.Cut()
+	icon_cache.Cut()
+	cache_size = 0
+	for (var/atom/atom)
+		atom.flags &= ~OVERLAY_QUEUED
+		CHECK_TICK
+
+
+/datum/controller/subsystem/overlays/Initialize(timeofday)
+	fire(FALSE, TRUE)
+
+
+/datum/controller/subsystem/overlays/stat_entry()
+	..("Queued Atoms: [queue.len], Cache Size: [cache_size]")
+
+
+/datum/controller/subsystem/overlays/fire(resumed, no_mc_tick)
+	var/count = 1
+	for (var/atom/atom as anything in queue)
+		++count
+		atom?.UpdateOverlays()
+		if (no_mc_tick)
+			CHECK_TICK
+		else if (MC_TICK_CHECK)
+			queue.Cut(count)
+			return
+	queue.Cut()
+
+
+/datum/controller/subsystem/overlays/proc/GetStateAppearance(icon, state)
+	var/list/subcache = state_cache[icon]
+	if (!subcache)
+		subcache = list()
+		state_cache[icon] = subcache
+	if (!subcache[state])
+		var/image/image = new (icon, null, state)
+		subcache[state] = image.appearance
+		++cache_size
+	return subcache[state]
+
+
+/datum/controller/subsystem/overlays/proc/GetIconAppearance(icon)
+	if (!icon_cache[icon])
+		var/image/image = new (icon)
+		icon_cache[icon] = image.appearance
+		++cache_size
+	return icon_cache[icon]
+
+
+/datum/controller/subsystem/overlays/proc/GetAppearanceList(atom/subject, list/sources)
+	if (!sources)
+		return list()
+	if (!islist(sources))
+		sources = list(sources)
+	var/list/result = list()
+	var/icon/icon = subject.icon
+	for (var/atom/entry as anything in sources)
+		if (!entry)
+			continue
+		else if (istext(entry))
+			result += GetStateAppearance(icon, entry)
+		else if (isicon(entry))
+			result += GetIconAppearance(entry)
+		else
+			if (isloc(entry))
+				if (entry.flags & OVERLAY_QUEUED)
+					entry.ImmediateOverlayUpdate()
+			if (!ispath(entry))
+				result += entry.appearance
+			else
+				var/image/image = entry
+				result += image.appearance

```

Comment:
replaces passing appearance through an image with just using it, still allows the questionable path behavior by giving it a clean image